### PR TITLE
a set of functions for the expression

### PR DESCRIPTION
**Pull request by ChristianBeilschmidt**

### CODE REVIEW

**Comment by ChristianBeilschmidt** [comment]
[Liste von Funktionen für Expression](https://trello.com/c/pnSh9Qj3/569-liste-von-funktionen-f%C3%BCr-expression)

File: operators/src/processing/expression/functions.rs [hunk 1]
```diff
@@ -0,0 +1,142 @@
+use proc_macro2::{Ident, TokenStream};
+use quote::quote;
+use std::{collections::HashMap, ops::RangeInclusive};
+
+pub(super) struct Function {
+    pub num_args: RangeInclusive<usize>,
+    pub token_fn: fn(usize, Ident) -> TokenStream,
```
[the remark this answers is not in the document]
Comment (ChristianBeilschmidt) [reply operators/src/processing/expression/functions.rs:7]: I don't think it costs that much. We could add a cache but we also create other token stream parts ad-hoc.

File: operators/src/processing/expression/functions.rs [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+lazy_static::lazy_static! {
+    pub(super) static ref FUNCTIONS: HashMap<&'static str, Function> = {
+        let mut functions = HashMap::new();
+
+        functions.insert("min", Function {
+            num_args: 2..=2,
```
[the remark this answers is not in the document]
Comment (ChristianBeilschmidt) [reply operators/src/processing/expression/functions.rs]: 719ef00

File: operators/src/processing/expression/parser.rs [hunk 3]
```diff
@@ -88,7 +89,7 @@ impl ExpressionParser {
         ExpressionAst::new(
             name.to_string().into(),
             self.parameters,
-            self.imports.borrow().iter().cloned().collect(),
+            self.functions.borrow().iter().cloned().collect(),
```
[the remark this answers is not in the document]
Comment (ChristianBeilschmidt) [reply operators/src/processing/expression/parser.rs]: It is not that easy since it is inside an `Rc`, but I got rid of the `clone`:

3ac21d5

**Review by michaelmattig — APPROVED** [review]

**Comment by michaelmattig** [comment]
bors merge

**Pull request merged**